tests/test_compute_funs.py

```diff
@@ -1688,3 +1690,82 @@ def test_surface_equilibrium_geometry():
                 rtol=3e-13,
                 atol=1e-13,
             )
+
+
+@pytest.mark.unit
+def test_parallel_grad():
+    """Test geometric and physical methods of computing parallel gradients agree."""
+    eq = get("W7-X")
+    eq.change_resolution(2, 2, 2, 4, 4, 4)
+    data = eq.compute(
+        [
+            "e_zeta|r,a",
+            "B",
+            "B^zeta",
+            "|B|_z|r,a",
+            "grad(|B|)",
+            "(|e_zeta|_z)|r,a",
+            "B^zeta_z|r,a",
+            "|B|",
+        ]
+    )
+    np.testing.assert_allclose(data["e_zeta|r,a"], (data["B"].T / data["B^zeta"]).T)
+    np.testing.assert_allclose(
+        data["|B|_z|r,a"], dot(data["grad(|B|)"], data["e_zeta|r,a"])
+    )
+    # FIXME: Don't know why below test fails.
```

Comment:
commenting for reference

Reply:
try for precise_QA and other examples, see which pass/fail

Reply:
Everything but dummy stellarator fails

Reply:
Ah that is bad then...

Reply:
any clues about this? Is there any pattern in how much it fails by for different configs?

Reply:
The test will pass for all stellarators on axis (rho=0) and near axis e.g. rho <= 1e-4. As rho increases, so does the error. At rho=1e-3, it fails with atol=3e-7, At rho-0.5, max absolute error is 0.1, max relative is 35. (These numbers are when the resolution is not reduced, full resolution linear grid).

FYI, the right hand side requires an additional radial derivative to compute the left hand side.

Reply:
resolved